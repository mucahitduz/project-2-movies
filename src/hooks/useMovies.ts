import { useState, useEffect, useCallback } from "react";
import { MovieService } from "../services/MovieService";

export function useMovies(query: string) {
  const [movies, setMovies] = useState([]);
  const [isLoading, setIsLoading] = useState(false);

  const fetchMovies = useCallback(async () => {
    try {
      setIsLoading(true);
      const movieData = await MovieService.fetchMovies(query);
      setMovies(movieData);
      setIsLoading(false);
    } catch (err) {
      console.error(err);
      setIsLoading(false);
    }
  }, [query]);

  useEffect(() => {
    fetchMovies();
  }, [query, fetchMovies]);

  return { movies, isLoading };
}
