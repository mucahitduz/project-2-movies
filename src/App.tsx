import Navbar from "./components/Navbar";
import Main from "./components/Main";
import { useState } from "react";
import SearchInput from "./components/SearchInput";
import SearchedMovies from "./components/SearchedMovies";
import MovieListBox from "./components/MovieListBox";
import WatchedSummary from "./components/WatchedSummary";
import WatchedMoviesList from "./components/WatchedMoviesList";
import Loader from "./components/Loader";
import SelectedMovie from "./components/SelectedMovie";
import { WatchedMovie } from "./types/watched-movie";
import { useMovies } from "./hooks/useMovies";

export default function App() {
  const [watched, setWatched] = useState<WatchedMovie[]>([]);
  const [selectedId, setSelectedId] = useState("");
  const [query, setQuery] = useState("");

  const { movies, isLoading, fetchMovies } = useMovies(query);

  function handleSelectMovie(id: string) {
    setSelectedId(id);
  }

  function handleCloseMovie() {
    setSelectedId("");
  }

  function handleAddWatched(movie: WatchedMovie) {
    setWatched((prevWatched) => [...prevWatched, movie]);
  }

  function handleDeleteWatchedMovie(id: string) {
    setWatched((prevWatched) =>
      prevWatched.filter((movie) => movie.imdbId !== id)
    );
  }

  function handleSearch() {
    fetchMovies();
  }

  return (
    <>
      <Navbar>
        <div className="logo">
          <span role="img">üçø</span>
          <h1>Movies</h1>
        </div>
        <SearchInput
          query={query}
          setQuery={setQuery}
          onSearch={handleSearch}
        />
        <p className="num-results">
          Found <strong>{movies?.length}</strong> results
        </p>
      </Navbar>

      <Main>
        <MovieListBox>
          {isLoading ? (
            <Loader />
          ) : (
            <SearchedMovies movies={movies} onSelectMovie={handleSelectMovie} />
          )}
        </MovieListBox>
        <MovieListBox>
          <>
            {selectedId ? (
              <SelectedMovie
                selectedId={selectedId}
                onCloseMovie={handleCloseMovie}
                onAddWatched={handleAddWatched}
                watched={watched}
              />
            ) : (
              <>
                <WatchedSummary watched={watched} />
                <WatchedMoviesList
                  watched={watched}
                  onDeleteWatchedMovie={handleDeleteWatchedMovie}
                />
              </>
            )}
          </>
        </MovieListBox>
      </Main>
    </>
  );
}
