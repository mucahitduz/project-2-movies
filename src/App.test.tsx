import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import App from "./App";
import { useMovies } from "./hooks/useMovies";
import "@testing-library/jest-dom";

jest.mock("./hooks/useMovies", () => ({
  useMovies: jest.fn(),
}));

describe("App Component", () => {
  it("should render Navbar and display logo", () => {
    (useMovies as jest.Mock).mockReturnValue({
      movies: [],
      isLoading: false,
      fetchMovies: jest.fn(),
    });

    render(<App />);

    expect(screen.getByRole("img")).toBeInTheDocument();
    expect(screen.getByRole("heading", { level: 1 })).toHaveTextContent(
      "Movies"
    );
  });

  it("should show the loader while movies are loading", () => {
    (useMovies as jest.Mock).mockReturnValue({
      movies: [],
      isLoading: true,
      fetchMovies: jest.fn(),
    });

    render(<App />);

    expect(screen.getByText(/Loading.../)).toBeInTheDocument();
  });

  it("should render searched movies when not loading", async () => {
    (useMovies as jest.Mock).mockReturnValue({
      movies: [
        { imdbID: "1", Title: "Movie 1", Year: "2023", Poster: "" },
        { imdbID: "2", Title: "Movie 2", Year: "2023", Poster: "" },
      ],
      isLoading: false,
      fetchMovies: jest.fn(),
    });

    render(<App />);

    // Movie 1 ve Movie 2'nin render edilmesini bekliyoruz
    await waitFor(() => {
      expect(screen.getByText("Movie 1")).toBeInTheDocument();
      expect(screen.getByText("Movie 2")).toBeInTheDocument();
    });
  });

  it("should match the snapshot", () => {
    (useMovies as jest.Mock).mockReturnValue({
      movies: [],
      isLoading: false,
      fetchMovies: jest.fn(),
    });

    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render selected movie details when a movie is selected", async () => {
    (useMovies as jest.Mock).mockReturnValue({
      movies: [{ imdbID: "1", Title: "Movie A", Year: "2023", Poster: "" }],
      isLoading: false,
      fetchMovies: jest.fn(),
    });

    render(<App />);

    // Movie A'yı tıklayalım
    fireEvent.click(screen.getByText("Movie A"));

    // Movie A'nın detaylarını doğru şekilde render ettiğinden emin olalım
    await waitFor(() => {
      expect(screen.getByText("Movie A")).toBeInTheDocument();
    });
  });
});
