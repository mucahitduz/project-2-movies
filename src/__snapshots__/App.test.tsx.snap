// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match the snapshot 1`] = `
<DocumentFragment>
  <nav
    class="nav-bar"
  >
    <div
      class="logo"
    >
      <span
        role="img"
      >
        üçø
      </span>
      <h1>
        Movies
      </h1>
    </div>
    <div>
      <input
        class="search"
        placeholder="Search movies..."
        type="text"
        value=""
      />
    </div>
    <p
      class="num-results"
    >
      Found 
      <strong>
        0
      </strong>
       results
    </p>
  </nav>
  <main
    class="main"
  >
    <div
      class="box"
    >
      <button
        class="btn-toggle"
      >
        ‚Äì
      </button>
      <ul
        class="list list-movies"
      />
    </div>
    <div
      class="box"
    >
      <button
        class="btn-toggle"
      >
        ‚Äì
      </button>
      <div
        class="summary"
      >
        <h2>
          Movies you watched
        </h2>
        <div>
          <p>
            <span>
              #Ô∏è‚É£
            </span>
            <span>
              0 movies
            </span>
          </p>
          <p>
            <span>
              ‚≠êÔ∏è
            </span>
            <span>
              0.00
            </span>
          </p>
          <p>
            <span>
              üåü
            </span>
            <span>
              0.00
            </span>
          </p>
          <p>
            <span>
              ‚è≥
            </span>
            <span>
              0 min
            </span>
          </p>
        </div>
      </div>
      <ul
        class="list"
      />
    </div>
  </main>
</DocumentFragment>
`;
